src/core/hle/service/ssl/ssl.cpp

```diff
@@ -80,21 +107,295 @@ class ISslConnection final : public ServiceFramework<ISslConnection> {
         // clang-format on
 
         RegisterHandlers(functions);
+
+        shared_data->connection_count++;
+    }
+
+    ~ISslConnection() {
+        shared_data_->connection_count--;
+        if (fd_to_close_.has_value()) {
+            s32 fd = *fd_to_close_;
+            if (!do_not_close_socket_) {
+                LOG_ERROR(Service_SSL,
+                          "do_not_close_socket was changed after setting socket; is this right?");
+            } else {
+                auto bsd = system.ServiceManager().GetService<Service::Sockets::BSD>("bsd:u");
+                if (bsd) {
+                    auto err = bsd->CloseImpl(fd);
+                    if (err != Service::Sockets::Errno::SUCCESS) {
+                        LOG_ERROR(Service_SSL, "failed to close duplicated socket: {}", err);
+                    }
+                }
+            }
+        }
     }
 
 private:
     SslVersion ssl_version;
+    std::shared_ptr<SslContextSharedData> shared_data_;
+    std::unique_ptr<SSLConnectionBackend> backend_;
+    std::optional<int> fd_to_close_;
+    bool do_not_close_socket_ = false;
+    bool get_server_cert_chain_ = false;
+    std::shared_ptr<Network::SocketBase> socket_;
+    bool did_set_host_name_ = false;
+    bool did_handshake_ = false;
+
+    ResultVal<s32> SetSocketDescriptorImpl(s32 fd) {
+        LOG_DEBUG(Service_SSL, "called, fd={}", fd);
+        ASSERT(!did_handshake_);
+        auto bsd = system.ServiceManager().GetService<Service::Sockets::BSD>("bsd:u");
+        ASSERT_OR_EXECUTE(bsd, { return ResultInternalError; });
+        s32 ret_fd;
+        // Based on https://switchbrew.org/wiki/SSL_services#SetSocketDescriptor
+        if (do_not_close_socket_) {
+            auto res = bsd->DuplicateSocketImpl(fd);
+            if (!res.has_value()) {
+                LOG_ERROR(Service_SSL, "failed to duplicate socket");
```

Comment:
```suggestion
                LOG_ERROR(Service_SSL, "Failed to duplicate socket");
```
maybe also log the fd here